File: desktop_client/engines/mock_workbook.py
"""
Python analogue of tools/sim/mockWorkbook.ts's minimal ExcelScript.Workbook
mock. Deliberately mirrors the same class shape (getWorksheet/getUsedRange/
getRangeByIndexes/setValues/clear) rather than a more "Pythonic" dict API,
so that porting an Office Script's main() body into a *_engine.py module is
a close, low-risk line-for-line translation instead of a re-design - the
translation itself is the highest-risk part of this port, so it is kept as
mechanical as possible.

Same seed/dump JSON shape as mockWorkbook.ts's constructor/dump(): a plain
{sheetName: [[cell, ...], ...]} dict, row 0 is the header row. This lets
tools/sim/compare_engines.py feed the exact same seed file to both the real
TypeScript engines (via tools/sim/run_e2e.ts) and this Python port.
"""
from __future__ import annotations

import re

CellValue = object  # str | int | float | bool
Row = list


def _parse_a1_start_row(a1: str) -> int:
    m = re.match(r"^[A-Z]+(\d+):", a1)
    if not m:
        raise ValueError(f'mock_workbook: cannot parse A1 range "{a1}"')
    return int(m.group(1)) - 1


class MockRange:
    def __init__(self, ws: "MockWorksheet", start_row: int, start_col: int, data: list[Row]):
        self.ws = ws
        self.start_row = start_row
        self.start_col = start_col
        self.data = data

    def get_values(self) -> list[Row]:
        return [list(r) for r in self.data]

    def get_row_count(self) -> int:
        return len(self.data)

    def set_values(self, values: list[Row]) -> None:
        for i, value_row in enumerate(values):
            target_row = self.start_row + i
            while len(self.ws.data) <= target_row:
                self.ws.data.append([])
            row = self.ws.data[target_row]
            for j, v in enumerate(value_row):
                col = self.start_col + j
                while len(row) <= col:
                    row.append("")
                row[col] = v

    def set_value(self, value) -> None:
        self.set_values([[value]])

    def clear(self, apply_to=None) -> None:
        # Matches mockWorkbook.ts: truncates from start_row onward. No cell-
        # formatting concept in this mock, so "contents" vs "all" are the same.
        del self.ws.data[self.start_row :]


class MockWorksheet:
    def __init__(self, initial: list[Row]):
        self.data: list[Row] = [list(r) for r in initial]

    def get_used_range(self) -> MockRange | None:
        if len(self.data) == 0:
            return None
        return MockRange(self, 0, 0, self.data)

    def get_range(self, a1: str) -> MockRange:
        start_row = _parse_a1_start_row(a1)
        return MockRange(self, start_row, 0, [])

    def get_range_by_indexes(self, start_row: int, start_col: int, num_rows: int, num_cols: int) -> MockRange:
        return MockRange(self, start_row, start_col, [])


class MockWorkbook:
    def __init__(self, seed: dict[str, list[Row]]):
        self.sheets: dict[str, MockWorksheet] = {name: MockWorksheet(rows) for name, rows in seed.items()}

    def get_worksheet(self, name: str) -> MockWorksheet:
        if name not in self.sheets:
            raise KeyError(f'mock_workbook: sheet "{name}" does not exist in the seed state')
        return self.sheets[name]

    def dump(self) -> dict[str, list[Row]]:
        return {name: [list(r) for r in ws.data] for name, ws in self.sheets.items()}

File: tools/sim/mockWorkbook.ts
// Minimal mock of the ExcelScript.Workbook API surface actually used by
// office-scripts/*.ts, backed by an in-memory array-of-arrays per sheet.
// Purpose: run the REAL compiled engine code (not a reimplementation)
// end-to-end against real production data, chaining engines in sequence so
// each one's writes are visible to the next - the closest verification
// possible without live Excel. See tools/sim/run_e2e.ts.

export type CellValue = string | number | boolean;
export type Row = CellValue[];

function parseA1StartRow(a1: string): number {
  // Only needs to handle patterns actually used in this codebase, e.g.
  // "A2:Q200000", "A1:F500", "A1:AM100000" - extracts the 1-indexed start
  // row from the first cell reference and converts to a 0-indexed row.
  const m = a1.match(/^[A-Z]+(\d+):/);
  if (!m) {
    throw new Error(`mockWorkbook: cannot parse A1 range "${a1}"`);
  }
  return parseInt(m[1], 10) - 1;
}

class MockRange {
  constructor(private ws: MockWorksheet, private startRow: number, private startCol: number, private data: Row[]) {}

  getValues(): Row[] {
    return this.data.map((r) => [...r]);
  }

  getRowCount(): number {
    return this.data.length;
  }

  setValues(values: Row[]): void {
    for (let i = 0; i < values.length; i++) {
      const targetRow = this.startRow + i;
      while (this.ws.data.length <= targetRow) {
        this.ws.data.push([]);
      }
      const row = this.ws.data[targetRow];
      for (let j = 0; j < values[i].length; j++) {
        row[this.startCol + j] = values[i][j];
      }
    }
  }

  setValue(value: CellValue): void {
    this.setValues([[value]]);
  }

  clear(): void {
    // Truncates from startRow onward (matches every clear() call site in
    // this codebase, which always clears "from row X to the end").
    this.ws.data.length = Math.min(this.ws.data.length, this.startRow);
  }
}

class MockWorksheet {
  data: Row[] = [];

  constructor(initial: Row[]) {
    this.data = initial.map((r) => [...r]);
  }

  getUsedRange(): MockRange | null {
    if (this.data.length === 0) {
      return null;
    }
    return new MockRange(this, 0, 0, this.data);
  }

  getRange(a1: string): MockRange {
    const startRow = parseA1StartRow(a1);
    // A "range to clear" doesn't need real data attached - clear() only
    // needs to know where to truncate from.
    return new MockRange(this, startRow, 0, []);
  }

  getRangeByIndexes(startRow: number, startCol: number, numRows: number, numCols: number): MockRange {
    return new MockRange(this, startRow, startCol, []);
  }
}

export class MockWorkbook {
  private sheets: { [name: string]: MockWorksheet } = {};

  constructor(seed: { [sheetName: string]: Row[] }) {
    for (const name of Object.keys(seed)) {
      this.sheets[name] = new MockWorksheet(seed[name]);
    }
  }

  getWorksheet(name: string): MockWorksheet {
    if (!this.sheets[name]) {
      throw new Error(`mockWorkbook: sheet "${name}" does not exist in the seed state`);
    }
    return this.sheets[name];
  }

  dump(): { [sheetName: string]: Row[] } {
    let out: { [sheetName: string]: Row[] } = {};
    for (const name of Object.keys(this.sheets)) {
      out[name] = this.sheets[name].data.map((r) => [...r]);
    }
    return out;
  }
}
